src/System.Linq.Expressions/src/System/Dynamic/ExpandoObject.cs

```diff
@@ -246,15 +247,14 @@ internal bool IsDeletedMember(int index)
         private ExpandoData PromoteClassCore(ExpandoClass oldClass, ExpandoClass newClass)
         {
             Debug.Assert(oldClass != newClass);
+            ContractUtils.AssertLockHeld(LockObject);
 
-            lock (LockObject)
+            if (_data.Class == oldClass)
```

Comment:
PromoteClassCore tries to do an "interlocked" promotion. - if we still have the oldClass, then promote, otherwise we lost the race so just return the witness. The caller may try again.

The assert is just to make sure we are not doing a no-op promotion. Every newClass must be truly new.